src/ucp/tag/rndv.c

```diff
@@ -283,21 +385,165 @@ static ucs_status_t
 ucp_rndv_ats_handler(void *arg, void *data, size_t length, void *desc)
 {
     ucp_reply_hdr_t *rep_hdr = data;
-    ucp_request_t *req = (ucp_request_t*) rep_hdr->reqptr;
+    ucp_request_t *sreq = (ucp_request_t*) rep_hdr->reqptr;
 
     /* dereg the original send request and set it to complete */
-    ucp_request_send_buffer_dereg(req, ucp_ep_get_rndv_data_lane(req->send.ep));
-    ucp_request_complete_send(req, UCS_OK);
+    ucp_request_send_buffer_dereg(sreq, ucp_ep_get_rndv_data_lane(sreq->send.ep));
+    ucp_request_complete_send(sreq, UCS_OK);
+
+    return UCS_OK;
+}
+
+static size_t ucp_rndv_pack_single_data(void *dest, void *arg)
+{
+    ucp_rndv_data_hdr_t *hdr = dest;
+    ucp_request_t *sreq = arg;
+    size_t length;
+
+    ucs_assert(sreq->send.state.offset == 0);
+
+    hdr->rreq_ptr = sreq->send.proto.rreq;
+    length = ucp_tag_pack_dt_copy(hdr + 1, sreq->send.buffer,
+                                  &sreq->send.state, sreq->send.length,
+                                  sreq->send.datatype);
+    ucs_assert(length == sreq->send.length);
+    return sizeof(*hdr) + length;
+}
+
+static size_t ucp_rndv_pack_multi_data(void *dest, void *arg)
+{
+    ucp_rndv_data_hdr_t *hdr = dest;
+    ucp_request_t *sreq = arg;
+    size_t length;
+
+    hdr->rreq_ptr = sreq->send.proto.rreq;
+    length        = ucp_ep_config(sreq->send.ep)->max_am_bcopy - sizeof(*hdr);
+
+    return sizeof(*hdr) + ucp_tag_pack_dt_copy(hdr + 1, sreq->send.buffer,
+                                               &sreq->send.state, length,
+                                               sreq->send.datatype);
+}
+
+static size_t ucp_rndv_pack_multi_data_last(void *dest, void *arg)
+{
+    ucp_rndv_data_hdr_t *hdr = dest;
+    ucp_request_t *sreq = arg;
+    size_t length;
+
+    hdr->rreq_ptr = sreq->send.proto.rreq;
+    length        = sreq->send.length - sreq->send.state.offset;
+
+    return sizeof(*hdr) + ucp_tag_pack_dt_copy(hdr + 1, sreq->send.buffer,
+                                               &sreq->send.state, length,
+                                               sreq->send.datatype);
+}
+
+static ucs_status_t ucp_rndv_progress_send(uct_pending_req_t *self)
+{
+    ucp_request_t *sreq = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_ep_t *ep = sreq->send.ep;
+    ucs_status_t status;
+
+    sreq->send.lane = ucp_ep_get_am_lane(ep);
+
+    if (sreq->send.length <= ucp_ep_config(ep)->max_am_bcopy - sizeof(ucp_rndv_data_hdr_t)) {
+        /* send a single bcopy message */
+        ucs_trace_req("Progress send request %p on am lane: %d. single message "
```

Comment:
pls remove logging from this function or use trace_data level. trace_req should happen O(1) times per request, not for each fragment